File: Java/0100_Same_Tree.java
import java.util.Stack;

// 2020-08-04 21:42:15

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        Stack<TreeNode> stack = new Stack<>();
        stack.push(p);
        stack.push(q);
        while (!stack.isEmpty()) {
            TreeNode left = stack.pop();
            TreeNode right = stack.pop();
            if (left == null && right == null) continue;
            if (left == null || right == null) return false;
            if (left.val != right.val) return false;
            stack.push(left.left);
            stack.push(right.left);
            stack.push(left.right);
            stack.push(right.right);
        }
        if (stack.isEmpty()) return true;
        return false;
    }
}

File: TypeScript/0100-same-tree.ts
// 2020-08-07 16:24:17

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isSameTree(p: TreeNode | null, q: TreeNode | null): boolean {
  const stack: StackNode[] = [
    [p, q]
  ];

  while (stack.length > 0) {
    const [left, right] = stack.pop() as StackNode;
    if (!left && !right) continue;
    if (!left || !right) return false;
    if (left.val == right.val) {
      stack.push([left.left, right.left]);
      stack.push([left.right, right.right]);
    } else {
      return false;
    }
  }

  return true;
};

type StackNode = [
  TreeNode | null,
  TreeNode | null
];